### PR TITLE
Fix NPE caused by txn commit race condition

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**:

==COMMIT_MSG==
Fixed a race condition where if a transaction was just committed by another thread, we may return null values from TransactionService.get.
==COMMIT_MSG==

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Added a unit test that failed without the first commit

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**: today

### CODE REVIEW

File: changelog/@unreleased/pr-5940.v2.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+type: fix
+fix:
+  description: Fixed a race condition where if a transaction was just committed by
+    another thread, we may return null values from TransactionService.get.
```
Comment (jeremyk-91) [inline changelog/@unreleased/pr-5940.v2.yml:4]: Let's probably say what this means to a user: users may observe that a transaction that had _definitively_ committed or aborted was still running. Retrying the transaction should work.

**Review by jeremyk-91 — COMMENTED** [review]
generally looks good, 1 comment

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java [hunk 2]
```diff
@@ -100,10 +101,9 @@ public void putUnlessExistsMultiple(Map<Long, Long> keyValues) throws KeyAlready
                                 + "was found in the KVS",
                         SafeArg.of("kvsValue", kvsValue),
                         SafeArg.of("stagingValue", currentValue));
-                continue;
+            } finally {
+                resultBuilder.put(startTs, commitTs);
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:105]: This seems a bit odd. If an exception is thrown, won't this appear to a caller that a value was committed, but in reality it has not?

Comment (gsheasby) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java:105]: If a CheckAndSetException is thrown here, it means the value was committed by another thread, and hence we must return it.

If other exceptions are thrown, then we will propagate the exception out of this method, and thus would not return the result anyway.

This is quite subtle, though, so I'll add a clarifying comment.

**Review by jeremyk-91 — APPROVED** [review]
👍

**Pull request merged**